src/interfaces/cloud.rs

```diff
@@ -0,0 +1,345 @@
+        self.upsert_entries_key = reduce_option("upsert entries", self.upsert_entries_key, index)?;
+        self.insert_chains_key = reduce_option("insert chains", self.insert_chains_key, index)?;
+
+        Ok(())
+    }
+}
+
+fn reduce_option(
+    debug_info: &str,
+    permission: Option<[u8; 16]>,
+    keep: bool,
+) -> Result<Option<[u8; 16]>, FindexErr> {
+    if let Some(permission) = permission {
+        if keep {
+            Ok(Some(permission))
+        } else {
+            Err(FindexErr::Other(format!(
+                "The token provided doesn't have the permission to {debug_info}"
+            )))
+        }
+    } else {
+        Ok(None)
+    }
+}
+
+enum Callback {
+    FetchEntries,
+    FetchChains,
+    UpsertEntries,
+    InsertChains,
+}
+
+impl FindexCloud {
+    pub fn new(token: String, base_url: Option<String>) -> Result<Self, FindexErr> {
+        Ok(FindexCloud {
+            token: Token::from_str(&token)?,
+            base_url,
+        })
+    }
+
+    async fn post(&self, callback: Callback, bytes: &[u8]) -> Result<Vec<u8>, FindexErr> {
+        let endpoint = match callback {
+            Callback::FetchEntries => "fetch_entries",
+            Callback::FetchChains => "fetch_chains",
+            Callback::UpsertEntries => "upsert_entries",
+            Callback::InsertChains => "insert_chains",
+        };
+
+        let key = match callback {
+            Callback::FetchEntries => self.token.fetch_entries_key,
+            Callback::FetchChains => self.token.fetch_chains_key,
+            Callback::UpsertEntries => self.token.upsert_entries_key,
+            Callback::InsertChains => self.token.insert_chains_key,
+        }
+        .ok_or(FindexErr::Other(format!(
+            "your key doesn't have the permission to call {endpoint}"
+        )))?;
+
+        let mut hasher = Kmac::v128(&key, &[]);
+        let mut output = [0u8; 32];
+        hasher.update(bytes);
+        hasher.finalize(&mut output);
+
+        let signature = hex::encode(output);
+
+        let url = format!(
+            "{}/indexes/{}/{endpoint}",
+            self.base_url
+                .as_deref()
+                .unwrap_or("https://findex.cosmian.com"),
+            self.token.index_id,
+        );
+
+        let client = reqwest::Client::new();
+        let res = client
+            .post(url)
+            .header("X-Findex-Cloud-Signature", &signature)
+            .body(bytes.to_vec())
+            .send()
+            .await
+            .map_err(|err| FindexErr::Other(format!("err {err}")))?;
+
+        Ok(res
+            .bytes()
+            .await
+            .map_err(|err| FindexErr::Other(format!("err2 {err}")))?
+            .to_vec())
+    }
+}
+
+impl FindexCallbacks<UID_LENGTH> for FindexCloud {
+    async fn progress(
+        &self,
+        _results: &HashMap<Keyword, HashSet<IndexedValue>>,
+    ) -> Result<bool, FindexErr> {
+        Ok(true)
+    }
```

Comment:
I agree with @ackRow: the `progress` callback should be removed from the `FindexCallbacks` trait and put back inside the `FindexSearch` trait as it is not managed the same way as the others.

This would complicate this MR so I think you can open an issue about it.

Reply:
Done in #24